internal/provider/restore_resource_test.go

```diff
@@ -0,0 +1,571 @@
+		clusterResourceName = "cockroach_cluster.test_cluster"
+	)
+	resource.Test(t, resource.TestCase{
+		IsUnitTest:               useMock,
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: getTestRestoreResourceConfig(clusterName, false /*includeRestore*/, restoreType, objects, opts, isAdvancedCluster),
+				Check: resource.ComposeTestCheckFunc(
+					func(s *terraform.State) error {
+						if !useMock {
+							clusterResource := s.RootModule().Resources["cockroach_cluster.test_cluster"]
+							clusterID := clusterResource.Primary.ID
+
+							t.Logf("Waiting for backups to be available for cluster %s...", clusterID)
+							return retry.RetryContext(ctx, 10*time.Minute,
+								waitForBackupReadyFunc(ctx, clusterID, service))
+						}
+						return nil
+					},
+				),
+			},
+			{
+				Config: getTestRestoreResourceConfig(clusterName, true /*includeRestore*/, restoreType, objects, opts, isAdvancedCluster),
+				Check:  testCheckRestoreResource(restoreResourceName, clusterResourceName, restoreType),
+			},
+			{
+				Config:      getTestRestoreResourceConfig(clusterName, true /*includeRestore*/, newRestoreType, objects, opts, isAdvancedCluster),
+				ExpectError: regexp.MustCompile("Immutable resource update attempted"),
+			},
+			{
+				ResourceName: restoreResourceName,
+				ImportState:  true,
+				ExpectError:  regexp.MustCompile("Import not supported"),
+			},
+		},
+	})
+}
+
+func getTestRestoreResourceConfig(
+	clusterName string,
+	includeRestore bool,
+	restoreType string,
+	objects []RestoreItem,
+	opts *RestoreOpts,
+	isAdvancedCluster bool,
+) string {
+	resources := fmt.Sprintf(`
+resource "cockroach_cluster" "test_cluster" {
+	name           = "%s"
+	cloud_provider = "GCP"
+	plan           = "STANDARD"
+	serverless = {
+		usage_limits = {
+			provisioned_virtual_cpus = 2
+		}
+		upgrade_type = "AUTOMATIC"
+	}
+	regions = [{
+		name: "us-central1"
+	}]
+	backup_config = {
+		enabled           = true
+		frequency_minutes = 5
+		retention_days    = 30
+	}
+}
+`, clusterName)
+
+	if isAdvancedCluster {
+		resources = fmt.Sprintf(`
+resource "cockroach_cluster" "test_cluster" {
+	name           = "%s"
+	cloud_provider = "GCP"
+	cockroach_version = "%s"
+	dedicated = {
+		storage_gib = 15
+		num_virtual_cpus = 4
+		cidr_range = "172.28.0.0/16"
+	}
+	regions = [{
+		name: "us-central1"
+		node_count: 1
+	}]
+}
+`, clusterName, latestClusterMajorVersion)
+	}
+
+	if includeRestore {
+		resources += fmt.Sprintf(`
+resource "cockroach_restore" "test_restore" {
+	destination_cluster_id = cockroach_cluster.test_cluster.id
+	type = "%s"
+	source_cluster_id = cockroach_cluster.test_cluster.id
+	%s
+	%s
+}
+`, restoreType, constructRestoreItems(objects), constructRestoreOpts(opts))
+	}
+	return resources
+}
+
+func constructRestoreItems(objects []RestoreItem) string {
+	var parts []string
```

Comment:
parts is a bit generic of a name.  How about something like `objectBlocks`.  Or another fine option would be to use a name like config and just append the the string to itself.

```
config := ""
for _, o := range objects {
     ....
     config += fmt.Sprintf(`objects {
     ... 
```